pysiaf/utils/tools2.py

```diff
@@ -0,0 +1,217 @@
+"""A collection of helper functions to support pysiaf calculations and functionalities
+
+Authors
+-------
+
+    - Colin Cox
+
+References
+----------
+
+
+
+"""
+
+import sys
+import numpy as np
+from math import *
+import copy
+import pylab as pl
+
+from astropy.table import Table
+
+from ..aperture import PRD_REQUIRED_ATTRIBUTES_ORDERED
+from .polynomial import poly, ShiftCoeffs, Rotate, FlipY, FlipX, rotate_coefficients, RotateCoeffs, triangle, invert
+from ..siaf import Siaf
+from pysiaf.aperture import DISTORTION_ATTRIBUTES
+
+def checkinv(A, B, C, D, order):
+    '''Check that forward and inverse transformations are consistent'''
+    [x,y] = np.mgrid[-200:201:200, -200:201:200]
+    v2 = poly(A, x, y, order)
+    v3 = poly(B, x, y, order)
+    x2 = poly(C, v2, v3, order)
+    y2 = poly(D, v2, v3, order)
+
+    print ('     x         y          V2        V3        xp        yp       dx        dy')
+    ######   -200.0000 -200.0000  -12.5849  -12.5939 -199.7371 -200.2741    0.2629   -0.2741
+
+    for i in range(3):
+        for j in range(3):
+            print(8*'%10.4f' %(x[i,j],y[i,j],v2[i,j],v3[i,j],x2[i,j],y2[i,j], x2[i,j]-x[i,j], y2[i,j]-y[i,j]))
+
+    # Summary
+    print ('\nShifts', (x2-x).mean(), (y2-y).mean())
+    print ('RMS {:10.2e} {:10.2e}'.format( (x2-x).std(), (y2-y).std() ))
+    print()
+
+    return
+
+def get_polynomial_coefficients(aperture, group):
+    '''Extract one of the four polynomial groups for an aperture
+    The group will be one of 'Sci2IdlX', 'Sci2IdlY', 'Idl2SciX', 'Idl2SciY' '''
+    coeffList = [getattr(aperture, s) for s in DISTORTION_ATTRIBUTES if group in s]
+    coefficients = np.array(coeffList, dtype=float)
+    return coefficients
+
+def matchV2V3(apName1, apName2):
+    """ Use the V2V3 from aperture named apName1 in aperture apName2, modifying XDetRef, YDetRef,
+    XSciRef YSciRef to match
+    Also  copy or shift the polynomial coefficients to reflect the new reference point origin
+    and for NIRCam recalculate angles. """
+
+    instruments = {'NRC':'NIRCam', 'NRS':'NIRSpec', 'FGS':'FGS', 'MIR':'MIRI', 'NIS':'NIRISS'}
+    insCode = apName1[:3]
+    ins = instruments[insCode]
+    print ('\n******************************************************')
+    print ('Instrument', ins)
+    siaf = Siaf(ins)
+    ap1 = siaf.apertures[apName1]
+    ap2 = siaf.apertures[apName2]
+    assert (ap2.AperType in ['FULLSCA', 'SUBARRAY', 'ROI']), "2nd aperture must be pixel-based"
+    order = ap1.Sci2IdlDeg
+    V2Ref1 = ap1.V2Ref
+    V3Ref1 = ap1.V3Ref
+    newV2Ref = V2Ref1 # in all cases
+    newV3Ref = V3Ref1
+    print ('Current Vref', ap2.V2Ref, ap2.V3Ref)
+    print ('Shift to    ', V2Ref1, V3Ref1)
+
+    # Detector and Science axes may go in opposite directions
+    ySign = cos(radians(ap2.DetSciYAngle))
```

Comment:
Using numpy, this would be
`y_sign = np.cos(np.deg2rad(aperture_2.DetSciYAngle))`